dynamiqs/utils/utils.py

```diff
@@ -536,7 +535,42 @@ def ket_overlap(x: Tensor, y: Tensor) -> Tensor:
     return (ket_to_bra(x) @ y).squeeze(-1).sum(-1)
 
 
-def ket_fidelity(x: Tensor, y: Tensor) -> Tensor:
+def fidelity(x: Tensor, y: Tensor) -> Tensor:
+    r"""Returns the fidelity of two states, kets or density matrix.
+
+    Warning:
+        This definition is different from `qutip.fidelity()` which uses the square root
+        fidelity $F_\text{qutip} = \sqrt{F}$.
+
+    Args:
+        x _(..., n, 1) or (..., n, n)_: First ket or density matrix.
+        y _(..., n, 1) or (..., n, n)_: Second ket or density matrix.
+
+    Returns:
+        _(...)_ Real-valued fidelity.
+
+    Examples:
+        >>> fock0 = dq.fock(3, 0)
+        >>> dq.fidelity(fock0, fock0)
+        tensor(1.)
+        >>> fock01 = 0.5 * (dq.ket_to_dm(dq.fock(3, 1)) + dq.ket_to_dm(dq.fock(3, 0)))
+        >>> dq.fidelity(fock01, fock01)
+        tensor(1.000)
+        >>> dq.fidelity(fock0, fock01)
+        tensor(0.500)
+    """
+
+    if isket(x) and isket(y):
+        return _ket_fidelity(x, y)
+    elif isket(x):
+        return _ket_dm_fidelity(x, y)
+    elif isket(y):
+        return _ket_dm_fidelity(y, x)
+    else:
+        return _dm_fidelity(x, y)
+
+
+def _ket_fidelity(x: Tensor, y: Tensor) -> Tensor:
```

Comment:
I think it's costly and useless to maintain clean docstrings for internal functions (developers are knowledgeable, they don't need as much info), let's remove them and shorten the internal doc to:

For `_ket_fidelity`:
```python
# returns the fidelity of two kets: |<x, y>|^2
# x: (..., n, 1), y: (..., n, 1) -> (...)
```

For `_dm_fidelity`:
```python
# returns the fidelity of two density matrices: Tr[sqrt(sqrt(x) @ y @ sqrt(x))]^2
# x: (..., n, n), y: (..., n, n) -> (...)
```

For `_ket_dm_fidelity`:
```python
# returns the fidelity of a ket `x` and a density matrix `y`: <x|y|x>
# x: (..., n, 1), y: (..., n, n) -> (...)
```